mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -300,6 +300,66 @@ class MatMulConverter final : public OpConversionPattern<tosa::MatMulOp> {
   }
 };
 
+class ReduceSumConverter final : public OpConversionPattern<tosa::ReduceSumOp> {
+public:
+  using OpConversionPattern<tosa::ReduceSumOp>::OpConversionPattern;
+
+  // This is a simple heuristic to calculate gridSize and blockSize
+  void calculateLaunchDims(int64_t numElements, int64_t num_cu,
+                           int64_t &gridSize, int64_t &blockSize,
+                           int64_t &elemPerThread) const {
+    // Currently fixing the blockSize to 256 for standalone reductions
+    blockSize = 256;
+
+    int64_t maxGridSize = rock::getMaxGridSize(num_cu);
+    gridSize = (numElements + blockSize - 1) / blockSize;
+    if (num_cu > 0) {
+      gridSize = std::max(gridSize, maxGridSize);
+    }
+
+    int64_t totalThreads = gridSize * blockSize;
+    elemPerThread = (numElements + totalThreads - 1) / totalThreads;
+  }
+
+  LogicalResult matchAndRewrite(tosa::ReduceSumOp op,
+                                tosa::ReduceSumOp::Adaptor adaptor,
+                                ConversionPatternRewriter &rw) const final {
+    Location loc = op->getLoc();
+    auto outputType = op.getType().cast<RankedTensorType>();
+    Value output =
+        rw.create<bufferization::AllocTensorOp>(loc, outputType, ValueRange{});
+
+    StringAttr arch;
+    Optional<uint32_t> num_cu;
+    rock::GemmFeatures features;
+    std::tie(arch, num_cu, features) = getArchAttributes(op);
+
+    int64_t gridSize;
+    int64_t blockSize;
+    int64_t elemPerThread;
+    int64_t numElements =
+        adaptor.getInput().getType().cast<ShapedType>().getNumElements();
+    calculateLaunchDims(numElements,
+                        num_cu.has_value() ? (int64_t)num_cu.value() : 0,
+                        gridSize, blockSize, elemPerThread);
+
+    rw.create<rock::ZeroInitKernelOp>(
```

Comment:
Please ensure that this goes into a separate `gpu.launch` from the `rock.reduce` call. Not doing that means race conditions.

Reply:
Right!

Would it better than a global sync here ?

Reply:
I see.

Reply:
(I mean, there _is_, and it has the form "allocate global memory, have each block inspect the total number of blocks we've been launched with, and do a fetch_add followed by a atomic spin wait for all the blocks to reach this point")

Reply:
Strictly saying, reduce op is reading the output buffer (read by atomic op).
Theoretically, execution layer should evaluate the dataflow analysis and set the dependency between the launch.
- dependency can be represented in async dialect but not sure we can lower to it.
- Assuming there is a layer performs the dataflow analysis, not sure if it handles WAW dependency from the given IR order, otherwise we may want to get the output buffer as an input argument as well.
@sjw do we have an execution-level pipeline which can evaluate dataflow among multiple kernels and run them?

Reply:
The thing is that creating a kernel operation (like the zero_init_kernel) doesn't actually create a kernel launch.

It's perfectly valid to, say, write
```
func.func @f(%a, %b, %c, %d, %e) attributes {kernel} {
    rock.gemm %c = %a * %b
    rock.gemm %e = %c * %d
}
```
but this will not ensure synchronization on `%c.

Similarly
````
func.func @f(%a, %b, %c) attributes {kernel} {
   rock.zero_init_kernel %c
  rock.gemm %c = %a * %b
}
launch @f
```

is absolutely not equivalent to
```
func.func @f(%a, %b, %c) attributes {kernel} {
  rock.zero_init_kernel %c
}
func.func @g(%a, %b, %c) attributes {kernel} {
  rock.gemm %c = %a * %b
}
launch @f 
launch @g 
```

Reply:
OK, I went back to the drawing board for this one.. 

From a kernel generation PoV what we want here is : 

```llvm
module @__xmodule_amdgcn_amd_amdhsa_gfx90a_srameccY_xnackY attributes {xmodel.arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", xmodel.module} {
    func.func private @test_reduce__part_0_0(%arg0: memref<2x3x1xf32>) attributes {kernel, original_func = @test_reduce__part_0, launch_func = @launch_test_reduce__part_0} {
      rock.zero_init_kernel %arg0 {arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", blockSize = 256 : i32, elemsPerThread = 1 : index, gridSize = 1 : i32} : memref<2x3x1xf32>
      return
    }
    func.func private @test_reduce__part_0_1(%arg0: memref<2x3x40xf32>, %arg1: memref<2x3x1xf32>) attributes {kernel, original_func = @test_reduce__part_0, launch_func = @launch_test_reduce__part_0} {
      rock.reduce  sum(%arg0, %arg1) {axis = 2 : index, blockSize = 256 : i32, gridSize = 1 : i32} : (memref<2x3x40xf32>, memref<2x3x1xf32>)
      return
    }
  }
}
```

Right ? 

Let me zoom out and share my thoughts on a possibility of a partitioning flow that gives access to rewriting the launches per op.

Currently we do the partitioning as follows : 

```llvm
module {
  func.func private @test_reduce__part_0(%arg0: tensor<2x3x40xf32> {func.read_access}) -> (tensor<2x3x1xf32> {func.write_access}) {
    %0 = "tosa.reduce_sum"(%arg0) {axis = 2 : i64} : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
    return %0 : tensor<2x3x1xf32>
  }
  func.func @test_reduce(%arg0: tensor<2x3x40xf32>) -> tensor<2x3x1xf32> {
    %token, %results = async.launch @test_reduce__part_0 (%arg0) : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
    async.await %token : !async.token
    return %results : tensor<2x3x1xf32>
  }
  module @__xmodule_amdgcn_amd_amdhsa_gfx90a_srameccY_xnackY attributes {xmodel.arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", xmodel.module} {
    func.func private @test_reduce__part_0(%arg0: tensor<2x3x40xf32> {func.read_access}) -> (tensor<2x3x1xf32> {func.write_access}) attributes {kernel, original_func = @test_reduce__part_0} {
      %0 = "tosa.reduce_sum"(%arg0) {axis = 2 : i64} : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
      return %0 : tensor<2x3x1xf32>
    }
  }
}
```

What if we do the partitioning as follows : 

```llvm
module {
  func.func private @test_reduce__part_0(%arg0: tensor<2x3x40xf32> {func.read_access}) -> (tensor<2x3x1xf32> {func.write_access}) {
    %0 = "tosa.reduce_sum"(%arg0) {axis = 2 : i64} : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
    return %0 : tensor<2x3x1xf32>
  }
  func.func private @launch_test_reduce__part_0(%arg0: tensor<2x3x40xf32> {func.read_access}) -> (tensor<2x3x1xf32> {func.write_access}) {
    %token, %results = async.launch @test_reduce__part_0 (%arg0) : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
    async.await %token : !async.token
    return %results : tensor<2x3x1xf32>
  }
  func.func @test_reduce(%arg0: tensor<2x3x40xf32>) -> tensor<2x3x1xf32> {
    %0 = call @test_reduce__part_0(%arg0) : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
    return %0 : tensor<2x3x1xf32>
  }
  module @__xmodule_amdgcn_amd_amdhsa_gfx90a_srameccY_xnackY attributes {xmodel.arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", xmodel.module} {
    func.func private @test_reduce__part_0(%arg0: tensor<2x3x40xf32> {func.read_access}) -> (tensor<2x3x1xf32> {func.write_access}) attributes {kernel, original_func = @test_reduce__part_0, launch_func = @launch_test_reduce__part_0} {
      %0 = "tosa.reduce_sum"(%arg0) {axis = 2 : i64} : (tensor<2x3x40xf32>) -> tensor<2x3x1xf32>
      return %0 : tensor<2x3x1xf32>
    }
  }
}
```

that could get lowered to as follows using the launch_func attr that would allows us to rewrite the launches as well :

```llvm
module {
  func.func private @test_reduce__part_0(%arg0: memref<2x3x40xf32> {func.read_access}, %arg1: memref<2x3x1xf32> {func.write_access}) {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = memref.alloc() {alignment = 128 : i64} : memref<2x3xf32>
    linalg.fill ins(%cst : f32) outs(%0 : memref<2x3xf32>)
    linalg.generic {indexing_maps = [#map0, #map1], iterator_types = ["parallel", "parallel", "reduction"]} ins(%arg0 : memref<2x3x40xf32>) outs(%0 : memref<2x3xf32>) {
    ^bb0(%arg2: f32, %arg3: f32):
      %2 = arith.addf %arg2, %arg3 : f32
      linalg.yield %2 : f32
    }
    %1 = memref.expand_shape %0 [[0], [1, 2]] : memref<2x3xf32> into memref<2x3x1xf32>
    memref.copy %1, %arg1 : memref<2x3x1xf32> to memref<2x3x1xf32>
    return
  }
  func.func private @launch_test_reduce__part_0(%arg0: memref<2x3x40xf32> {func.read_access}, %arg1: memref<2x3x1xf32> {func.write_access}) {
    %token, %results = async.launch @test_reduce__part_0_0 (%arg0) : (memref<2x3x40xf32) -> ()
    async.await %token : !async.token
    %token, %results = async.launch @test_reduce__part_0_1 (%arg0, %arg1) : (memref<2x3x40xf32, memref<2x3x1xf32>) -> ()
    async.await %token : !async.token
    return %results : tensor<2x3x1xf32>
  }
  func.func @test_reduce(%arg0: memref<2x3x40xf32>, %arg1: memref<2x3x1xf32>) {
    %0 = call @launch_test_reduce__part_0(%arg0,  %arg1) : (memref<2x3x40xf32>, memref<2x3x1xf32>) -> ()
    return %0 : tensor<2x3x1xf32>
  }
  module @__xmodule_amdgcn_amd_amdhsa_gfx90a_srameccY_xnackY attributes {xmodel.arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", xmodel.module} {
    func.func private @test_reduce__part_0_0(%arg0: memref<2x3x1xf32>) attributes {kernel, original_func = @test_reduce__part_0, launch_func = @launch_test_reduce__part_0} {
      rock.zero_init_kernel %arg0 {arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack+", blockSize = 256 : i32, elemsPerThread = 1 : index, gridSize = 1 : i32} : memref<2x3x1xf32>
      return
    }
    func.func private @test_reduce__part_0_1(%arg0: memref<2x3x40xf32>, %arg1: memref<2x3x1xf32>) attributes {kernel, original_func = @test_reduce__part_0, launch_func = @launch_test_reduce__part_0} {
      rock.reduce  sum(%arg0, %arg1) {axis = 2 : index, blockSize = 256 : i32, gridSize = 1 : i32} : (memref<2x3x40xf32>, memref<2x3x1xf32>)
      return
    }
  }
}

I think it would not stop us serving the existing clients if they dont want to call the launch_func -- we could just export metadata about the kernels in the way they want to call the kernels.

cc : @sjw36 too.

Reply:
Pushed out tosa to rock conversions for now...